File: 0001-0500/0041.First Missing Positive.py
"""
Your runtime beats 60.19% of python submissions.
"""
class Solution(object):
    def firstMissingPositive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums=map(lambda x:0 if x<=0 or x>len(nums) else x,nums)
        m = len(nums)+1
        for x in xrange(len(nums)):
            if nums[x]<=0:
                continue
            else:
                nums[nums[x]%m-1]=nums[nums[x]%m-1]%m+m
        for x in xrange(len(nums)):
            if nums[x]<=len(nums):
                return x+1
        return len(nums)+1
                

File: 0001-0500/0041.First Missing Positive.rb
# @param {Integer[]} nums
# @return {Integer}
def first_missing_positive(nums)
    size = nums.size
    mod = size << 1
    for i in 0...nums.size
        if nums[i] <= 0 || nums[i] > size
            nums[i] = 0
        end
    end
    for i in 0...nums.size
        if !(nums[i] % mod).zero?
            nums[nums[i] % mod - 1] += mod
        end
    end
    for i in 0...nums.size
        if nums[i] < mod
            return i + 1
        end
    end
    size + 1
end

